elyra/pipeline/kfp/kfp_metadata.py

```diff
@@ -33,15 +33,58 @@ def on_load(self, **kwargs: Any) -> None:
             # Inject auth_type property for metadata persisted using Elyra < 3.3:
             # - api_username and api_password present -> use DEX Legacy
             # - otherwise -> use no authentication type
-            if self.metadata.get('api_username') is None or\
-               len(self.metadata.get('api_username').strip()) == 0 or\
-               self.metadata.get('api_password') is None or\
-               len(self.metadata.get('api_password').strip()) == 0:
+            if len(self.metadata.get('api_username', '').strip()) == 0 or\
+               len(self.metadata.get('api_password', '').strip()) == 0:
                 self.metadata['auth_type'] = SupportedAuthProviders.NO_AUTHENTICATION.name
             else:
                 self.metadata['auth_type'] = SupportedAuthProviders.DEX_LEGACY.name
 
+        if self.metadata.get('cos_auth_type') is None:
+            # Inject cos_auth_type property for metadata persisted using Elyra < 3.4:
+            # - cos_username and cos_password must be present
+            # - cos_secret may be present (above statement also applies in this case)
+            if self.metadata.get('cos_username') and\
+               self.metadata.get('cos_password'):
+                if len(self.metadata.get('cos_secret', '').strip()) == 0:
+                    self.metadata['cos_auth_type'] = 'USER_CREDENTIALS'
+                else:
+                    self.metadata['cos_auth_type'] = 'KUBERNETES_SECRET'
+
             # save changes
             MetadataManager(schemaspace="runtimes").update(self.name, self, for_migration=True)
 
         return None
+
+    def pre_save(self, **kwargs: Any) -> None:
+        """
+        This method enforces conditional constraints related to
+        COS authentication type properties.
+        TODO: Remove after https://github.com/elyra-ai/elyra/issues/2338
+        was resolved.
+        """
+        super().pre_save(**kwargs)
+
+        if self.metadata.get('cos_auth_type') is None:
+            # nothing to do
+            return
+
+        if self.metadata['cos_auth_type'] == 'USER_CREDENTIALS':
+            if len(self.metadata.get('cos_username', '').strip()) == 0 or\
+               len(self.metadata.get('cos_password', '').strip()) == 0:
+                raise ValueError('A username and password are required '
+                                 'for the selected Object Storage authentication type.')
+            if len(self.metadata.get('cos_secret', '').strip()) > 0:
+                raise ValueError('Kubernetes secrets are not supported '
+                                 'for the selected Object Storage authentication type.')
```

Comment:
```suggestion
                raise ValueError('Kubernetes secrets are not supported '
                                 'for the selected Object Storage authentication type.'
                                 'Please clear the "Cloud Object Storage Credentials Secret" field.' )
```
Can we add an extra sentence here telling the user to clear the cos secret parameter in the config?
![image](https://user-images.githubusercontent.com/13340013/146068783-23955a14-7d03-40d0-8e72-77db779cce7c.png)

Reply:
just makes it quicker for someone to figure out what to do if they keep clicking and it wont save. either this or have the UI clear the field automatically after selecting a different cos auth type, but thats def more work

Reply:
This is exactly what I was getting at with this https://github.com/elyra-ai/elyra/pull/2354#discussion_r767062791.  If `cos_secret` isn't used by the `auth_type` why enforce it to not be set?  (Same goes for `AWS_IAM_ROLES_FOR_SERVICE_ACCOUNTS` (below) by also including `username` and `password` in the set of tolerated values.)

Reply:
Sorry, somehow the referenced comments doesn't open up for me. I get tis preview but can see the remainder of your comment:
![image](https://user-images.githubusercontent.com/13068832/146094591-cf4fd81e-04d7-4be2-94db-9a6917eeddfe.png)

> If cos_secret isn't used by the auth_type why enforce it to not be set?

My motivation for taking this approach was as follows:
 - Ensure compatibility and consistency - the saved schema instance must not include any properties that are not applicable. This approach should mimic the behavior one will observe (I am assuming), once conditional properties are supported: the stored schema instance file only includes relevant properties to avoid the need for interpretation.
 - I initially considered removing conditional properties that are not applicable in `pre_save()` to accomplish the above. However, doing so would lead to two potential usability issues: 
   1. When the user enters problematic values and no error is raised they might infer that these values are applicable, even though they are not. With the current limitation the UI already leads to confusion and this would add to it.
   1. When the user re-opens the RTC  those non-applicable property values would be "gone". A user _might_ come to believe that this behavior is a bug because they entered values and the save operation completed without an error.

Reply:
I do agree that changing the message wording to include an action would be helpful.

Reply:
A potential mitigation option for "[...] those non-applicable property values would be 'gone'. A user might come to believe that this behavior is a bug" might be to update the property description to explicitly state something like

```
This property is required for authentication type 'XYZ'. It does not apply to other authentication types. 
```

Reply:
> Sorry, somehow the referenced comments doesn't open up for me. I get tis preview but can see the remainder of your comment:

No worries - it's in one of the resolved comments above.

Yeah, it dawned on me that the UI for this is probably an issue since probably doesn't "hide" non-applicable properties.  I think this is going to be a usability problem until we have better support for this kind of thing - so whatever you want to do is okay with me.  I just figured ignoring unused properties is easier, allows users the ability to switch between types, and is more in line with how the server code will behave when `oneOf` support is present.

Will there be any hint to the user which fields are required per authentication type?

Reply:
> ignoring unused properties is easier, 

True, but it also introduces unpredictability with respect to which properties might be present in the persisted runtime metadata file. I am trying to avoid that. For example, if we were to take this approach, there are several property permutations one could find in those files for  the `AWS_IAM_ROLES_FOR_SERVICE_ACCOUNTS` type, depending on what the user decided to enter:
 - no username, no password, no secret
 - username, no password, no secret, ...
 - ...
 - username, password, secret
 
Of all the permutations only one would be semantically valid.

> allows users the ability to switch between types, 

While in edit mode that will still be possible with the delivery of this PR, but I don't know to what extend the UI will support something like that in the future. For illustrative purposes (I'm not trying to document any expectations here)  if a user chooses `USER_CREDENTIALS` and enters uid + password, then switches to the `AWS_IAM_ROLES_FOR_SERVICE_ACCOUNTS` auth type - which has no properties, and then switches to the `KUBERNETES_SECRET` auth type. Are the uid + password values supposed to be pre-populated?

> Will there be any hint to the user which fields are required per authentication type?

We've updated the RTC documentation

![image](https://user-images.githubusercontent.com/13068832/146258819-bfacb454-2024-401a-aebe-57a89bd8dce9.png)

and I am about to extend the property descriptions that are exposed as tool tips:

![image](https://user-images.githubusercontent.com/13068832/146258696-a21ac471-5f57-4ea2-9c16-fee7e69c7c51.png)

Reply:
Understood, thanks.